chat_ui/src/main/kotlin/io.skygear.plugins.chat.ui/ConversationFragment.kt

```diff
@@ -586,35 +626,65 @@ open class ConversationFragment() :
-            addMessageToBottom(message, Uri.parse("file://" + voiceRecordingFileName))
 
+            this.fragmentMessageSentListener?.onBeforeMessageSent(this, message)
+            addMessageToBottom(message, Uri.parse("file://" + voiceRecordingFileName))
             this.skygearChat?.addMessage(message, conv, object : SaveCallback<ChatMessage> {
                 override fun onSuccess(chatMsg: ChatMessage?) {
                     voiceRecordingFile.delete()
+                    callMessageSentSuccessListener(message)
                 }
 
                 override fun onFail(error: Error) {
                     Log.e(
                             ConversationFragment.TAG,
                             "Failed to send voice message: ${error.message}"
                     )
+                    this@ConversationFragment.fragmentMessageSentListener?.onMessageSentFailed(this@ConversationFragment, message, error)
                 }
             })
         }
     }
 
+    fun callMessageSentSuccessListener(chatMsg: ChatMessage?) {
+        chatMsg?.let {
+            this@ConversationFragment.fragmentMessageSentListener?.
+                    onMessageSentSuccess(this@ConversationFragment, chatMsg)
+        } ?: run {
+            this@ConversationFragment.fragmentMessageSentListener?.
+                    onMessageSentFailed(this@ConversationFragment,
+                            null, Error("No message sent"))
+        }
+    }
+
+
+    fun callMessageFetchSuccessListener(chatMsgs: List<ChatMessage> ?, isCached: Boolean) {
+        chatMsgs?.let {
+            this@ConversationFragment.fragmentMessageFetchListener?.onMessageFetchSuccess(this@ConversationFragment, it, isCached)
+        } ?: run {
+            this@ConversationFragment.fragmentMessageFetchListener?.
+                    onMessageFetchFailed(this@ConversationFragment,
+                            Error("No message fetched"))
+        }
+    }
+
     fun onSendMessage(input: String): Boolean {
+
         this.conversation?.let { conv ->
             val message = ChatMessage()
             message.body = input.trim()
+            this.fragmentMessageSentListener?.onBeforeMessageSent(this, message)
             this.addMessageToBottom(message)
             this.skygearChat?.addMessage(message, conv, object : SaveCallback<ChatMessage> {
                 override fun onSuccess(msg: io.skygear.plugins.chat.Message?) {
+                    callMessageSentSuccessListener(message)
                     msg?.let { this@ConversationFragment.updateMessage(msg) }
                 }
 
                 override fun onFail(error: Error) {
-                    this@ConversationFragment.updateMessage(message, error)
+                    this@ConversationFragment.fragmentMessageSentListener?.
+                            onMessageSentFailed(this@ConversationFragment, message, error)
+                    message?.let { this@ConversationFragment.updateMessage(message) }
```

Comment:
@howawong, now when we send a message in no network, the message should turn to failed immediately. This feature is broken in your rebased branch. Is it because you missed the error in this `updateMessage` calling?